### PR TITLE
hashcat: 2.0.0 -> 3.0.0

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Just a package update.
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

OpenCL is now a requirement.

If you see a `ERROR: clGetPlatformIDs() : -1001 : CL_UNKNOWN_ERROR` error it's because `/etc/OpenCL/vendors` is missing.

### CODE REVIEW

File: pkgs/development/libraries/opencl-icd/default.nix [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
     url = "https://forge.imag.fr/frs/download.php/716/ocl-icd-${version}.tar.gz";
     sha256 = "1rgaixwnxmrq2aq4kcdvs0yx7i6krakarya9vqs7qwsv5hzc32hc";
   };
 
+  buildInputs = [ ruby opencl-headers ];
+
+  configureFlags = [ "--enable-official-khronos-headers" ];
+
+  meta = with stdenv.lib; {
+    description = "this free ICD Loader can load any (free or non free) ICD.";
```
Comment (bjornfor) [inline pkgs/development/libraries/opencl-icd/default.nix]: Nitpick: "this free" => "Free" and drop the trailing dot.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -8679,6 +8679,8 @@ in
 
   opencl-icd = callPackage ../development/libraries/opencl-icd { };
 
+  opencl-intel = callPackage ../development/libraries/opencl-intel { };
```
Comment (bjornfor) [inline pkgs/top-level/all-packages.nix]: What's the opencl-intel attribute doing in this commit?

**Comment by bjornfor** [comment]
Nice, it seems they finally merged hashcat and oclHashcat!

File: pkgs/development/libraries/opencl-icd/default.nix (hunk 1 above)
Comment (zimbatm) [inline pkgs/development/libraries/opencl-icd/default.nix]: Fixed

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (zimbatm) [inline pkgs/top-level/all-packages.nix]: Fixed, leftover from me trying to make OpenCL work

**Comment by zimbatm** [comment]
Yup but OpenCL is now a requirement. Is OpenCL working on your computer?

**Comment by bjornfor** [comment]
Don't know, haven't tested OpenCL before. So it's not just "nix-env -i hashcat && hashcat" anymore?

**Comment by zimbatm** [comment]
No you need a working OpenCL installation, meaning you need a .icd file in `/etc/OpenCL/vendors`.

Right now I think only Nvidia boxes or non-nixos boxes might work: https://github.com/zimbatm/nixpkgs/blob/hashcat-3.0.0/nixos/modules/hardware/video/nvidia.nix#L65 . That's why I was trying to add the intel sdk ^_^

Another thing I tried is adding this to my configuration.nix:
```
environment.etc."OpenCL/vendors/intel-beignet.icd".source =
    "${pkgs.beignet}/etc/OpenCL/vendors/intel-beignet.icd";
```
Which gets me past the `ERROR: clGetPlatformIDs() : -1001 : CL_UNKNOWN_ERROR` but gives me something else instead:
```
 Device #1: Kernel m00000_a0.4251ce2a.kernel not found in cache! Building may take a while...
ASSERTION FAILED: Fail to compile kernel, may need to increase reserved registers for spilling.
  at file /tmp/nix-build-beignet-1.1.2.drv-0/Beignet-1.1.2-Source/backend/src/backend/gen_program.cpp, function virtual gbe::Kernel* gbe::GenProgram::compileKernel(const gbe::ir::Unit&, const string&, bool), line 202
Trace/breakpoint trap
```

**Comment by bjornfor** [comment]
Hm, so hashcat will be Nvidia-only on NixOS when this is merged? I don't have Nvidia GPU (at home), and I guess there are others in my situation, is this enough of a regression to either (1) first make OpenCL work for Intel (and AMD) or (2) keep hashcat 2.0 around (in addition to 3.0)?

**Comment by zimbatm** [comment]
Yeah maybe keeping hashcat 2.x would be a good idea until the OpenCL support in NixOS gets better

**Comment by grahamc** [comment]
Hey @zimbatm, are you still interested in getting this merged?

**Comment by zimbatm** [comment]
@grahamc not really, feel free to close or take over if you want

**Comment by grahamc** [comment]
For people coming later, this worked for me on a Dell 9560:

```
hardware.opengl.extraPackages = [ pkgs.beignet ];
```

**Pull request closed without merging**